onnxruntime/core/providers/cuda/tensor/transpose.cc

```diff
@@ -92,20 +92,76 @@ Status Transpose::DoTranspose(const Transpose& kernel,
   const std::vector<int64_t>& output_dims = output.Shape().GetDims();
 
   auto rank = static_cast<int32_t>(input_dims.size());
-  TensorPitches original_input_strides(input_dims);
-  TensorPitches original_output_strides(output_dims);
 
-  TArray<int64_t> input_strides(rank);
-  for (auto i = 0; i < rank; i++) {
-    input_strides[i] = original_input_strides[permutations[i]];
-  }
-  TArray<fast_divmod> output_strides(rank);
-  for (auto i = 0; i < rank; i++) {
-    output_strides[i] = fast_divmod(gsl::narrow_cast<int>(original_output_strides[i]));
+  // flatten the adjacent dimensions which are contiguous
+  // for example: permutations[0, 2, 3, 1] -> [0, 2, 1], permutations[0, 3, 1, 2] -> [0, 2, 1]
+  auto new_rank = rank;
+  std::vector<size_t> new_permutations(permutations);
```

Comment:
>new_permutations [](start = 22, length = 16)

my understanding is that `new_permutations`, `new_input_dims`, and `new_output_dims` will have `new_rank` useful elements (maybe i'm wrong). if so, it would be clearer to actually have the containers be that size, instead of having unused stuff at the end.

Reply:
But new_rank is unknown upfront. It will be decreased if there is adjacent dimensions can be flattened.

Reply:
how about using std::vector::resize() after new_rank is determined?
and later on these containers are used, for example: `TensorPitches new_input_strides(new_input_dims);`
does it make sense for new_input_dims to have the length of the original input_dims at that point?